### PR TITLE
context: Added fallback devices

### LINKED ISSUE #5681: Device permission compatibility
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

Related issues #4405 (PR #5620) #5524 #5481

We want to add permissions for device categories.

As of now we have `dri`, `kvm`, `shm`, `all`, and the recently added `input` (see #5481). The latter is why I'm filing this.

Device `all` is used in a lot of situations when needed access to input devices (game controllers), USB (see #5620 for that part) etc.

`input` was added in #5481 but this mean that using `--device=input` is only when flatpak 1.15.6 or later, which seriously limit its use. The only way to have this work out of the box is to add `--device=all` which make the case of using `--device=input` moot.

Solution:

We should have a mechanism to allow the transition. Here is the proposal:

For compatibility purpose, the packager will indicate `device=input`, ignored by old flatpaks, `device=all` understood by old flatpak, and `device=not-all,input` to indicate newer flatpak that `all` should be ignored for `input`.

Later, we add support for `usb`, then we would indicate `device=not-all,usb,input` to disable the device `all` for both `input` and the just added `usb`.

The importan part is the forward looking mechanism. Support for `device=not-all` is implemented the following way:

`device=not-all` always followed by comma separated devices. Example: `device=not-all,input,usb`. The interpretation is: if for all the devices specified if flatpak knows about the device and it is specified the device=all is deselected.

The behaviour of `dri`, `kvm`, `shm` is unchanged, BUT sepecifying `--device=not-all` will require to specify them explicitely.

Example:

flatpak < 1.15.6 (most distro)

```
--device=all
--device=input
--device=not-all,input
```

Device: all ON, as it doesn't recognize anything but `all`.

flatpak know `input` but not `usb`:

```
--device=all
--device=input
--device=not-all,input
```

Device: all OFF

```
--device=all
--device=input
--device=not-all,input,usb
```

Device: all OFF, because `usb` is not specified.

```
--device=all
--device=input
--device=usb
--device=not-all,input,usb
```

Device: all ON, because it doesn't know about USB devices so it needs to let them all.

flatpak know `input` and `usb`:

```
--device=all
--device=input
--device=usb
--device=not-all,input,usb
```

Device: all OFF. This is the forward compatibility.

Using `--device` should allow to not require to add anything to `flatpak-builder`.

While the options are presented as `finish-args` from the manifest the metadata form would luke this:

```
[Context]
devices=all;input;usb;not-all,input,usb;
```

About the syntax

The choice of comma `,` as a separator fits as the keyfile uses the semi-colon `;`. So there is no conflict. It's a free form string because it allow just ignoring it if the flatpak version is too old.

The choice of `not-all` seems to be clear. Someone suggested `not-all-if`. It can be considered if clearer.

### PR DESCRIPTION
**Pull request by hfiguiere**
This provide compatibility to disable all devices when there is new device option

Fixes #5681

- [x] Need to sort out the documentation - https://github.com/flatpak/flatpak-docs/pull/478
- [x] Validate with flatpak-builder

### CODE REVIEW

File: tests/test-exports.c [hunk 1]
```diff
@@ -397,6 +404,7 @@ test_full_context (void)
                      (GCompareDataFunc) flatpak_strcmp0_ptr, NULL);
   i = 0;
   g_assert_cmpstr (strv[i++], ==, "all");
+  g_assert_cmpstr (strv[i++], ==, "all,input");
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply tests/test-exports.c:407]: uh oh, yes it should be. will fix this.

File: common/flatpak-context.c [hunk 2]
```diff
@@ -296,32 +300,202 @@ flatpak_context_sockets_to_args (FlatpakContextSockets sockets,
...
+
+  state->devices |= fallback;
+}
+
+/*
+ * We will remove `all` if it is in any fallback, AND said fallback
+ * isn't necessary.
+ */
+FlatpakContextDevices
+flatapk_context_devices_with_fallback (FlatpakContext  *context)
```
[the remark this answers is not in the document]
Comment (hfiguiere) [reply common/flatpak-context.c:330]: ouch. and cut&paste spread it all over.

File: doc/devices.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+`kvm`: exposes KVM subsystem.
+
+`shm`: exposes the shared memory device. This is the only one not
+exposed with `all`.
+
+`input`: exposes input devices, include mice, game controllers and
+events. This is not necessary for mouse a keyboard input, but it is
+for direct device access or game controllers.
+
+`all` exposes everyting. Accept `shm`.
```
Comment (nbenitez) [inline doc/devices.md:27]: "Except"

**Comment by hfiguiere** [comment]
> Maybe `--device=input:or-fallback` or `--device=input:fallback`? Or `--device=fallback:input,all` as I suggested before? Or something along those lines?

I went for `fallback:`. Indeed while more verbose it seems that it will be less confusing.

**Comment by hfiguiere** [comment]
Draft of the documentation https://github.com/flatpak/flatpak-docs/pull/478

**Comment by hfiguiere** [comment]
So I tested with flatpak-builder and the only thing that is neded is just an up to date version of flatpak.

I think this is really ready for review.

I have broadened the tests.

I will make the USB support rely on this.

**Comment by swick** [comment]
So, turns out I'm really not a fan of this. Can we maybe consider something along the lines of this:

`--device all --nodevice-if "all:has-input-device" --nodevice-if "all:has-usb-portal" --device input --nodevice-if "input:has-wayland-inputfd"`

`flatpak override --nodevice-if "all:never"

Which would have the benefit that the condition on which to enable some device permission can be arbitrary and is not just tied to another device permission?

**Comment by swick** [comment]
I also think this doesn't work properly with the different priorities. I.e. if you have an override `--device=fallback:input,all` and then at the commandline `--device=all`, this might still remove the `--device=all` permission.

**Comment by hfiguiere** [comment]
> So, turns out I'm really not a fan of this. Can we maybe consider something along the lines of this:
> 
> `--device=all --device=input --nodevice-if="all:has-input-device" --nodevice-if="all:has-usb-portal" --nodevice-if="input:has-wayland-inputfd"`
> 
> Which would have the benefit that the condition on which to enable some device permission can be arbitrary and is not just tied to another device permission?

I think it is way too complicated with way too little benefit.

We are not trying to detect runtime. There is no wayland involved here either. `--device=input` only exposes `/dev/input`. This is just to handle the use of flatpak with tighter device permissions. Because today if a flatpak has `--device=input`, it will never work properly with any stable release (1.14.x) of flatpak. Given that some distro will still be in use in 5 years, it's important.

As for USB, which is a different PR and still WiP we can add the detection of the portal in flatpak to decide whether to enable or not depending on the conditions.

**Comment by swick** [comment]
What this MR does is make it possible to remove one device permission when another device permission is supported. This isn't really that useful IMO.

Yes, remove "all" when an app can use "input" is useful. The replacement for "all" when the app needed it for usb access however isn't the usb device permission, but the usb portal. Similarly, with wayland inputfd we might be able to replace "all" and "input" permissions.

The key point here is that an existing device permission might get replaced not just by a more constrained device permission but by a portal, wayland protocol, pipewire, ...

**Comment by hfiguiere** [comment]
> What this MR does is make it possible to remove one device permission when another device permission is supported. This isn't really that useful IMO.

What this MR does it allow using the `input` device permission on newer versions of flatpak while allowing to continue to use application that need this device on older version of flatpak. The one that ship with most distribution used by people.
 
> Yes, remove "all" when an app can use "input" is useful. The replacement for "all" when the app needed it for usb access however isn't the usb device permission, but the usb portal. 

There is no USB portal nor permission. There is no reason to specify `input` if you also need `all` because use need to use USB devices.

> Similarly, with wayland inputfd we might be able to replace "all" and "input" permissions.

what is the "wayland inputfd" you keep talking about? So far there is no such permission. And I can't see how it solve any of the problem here.

> The key point here is that an existing device permission might get replaced not just by a more constrained device permission but by a portal, wayland protocol, pipewire, ...

This is why `--socket=pulseaudio` expose `/dev/snd`. Because the thinking was the "we'll add some sort of other permission later". And it never happened. (work on the Audio portal has been postponed).

All of this assume that application WILL be adpated to use the portals when they exist, and trust me, I doubt most will.

**Comment by hfiguiere** [comment]
> I also think this doesn't work properly with the different priorities. I.e. if you have an override `--device=fallback:input,all` and then at the commandline `--device=all`, this might still remove the `--device=all` permission.

There is no priority. `--nodevice=fallback:input,all` will remove the fallback and the `--device=all` already present will work.

**Comment by hfiguiere** [comment]
The latest addition adds a working `--nodevice` for the fallback, and testing for it.

**Comment by hfiguiere** [comment]
that would be great to get this in before `input` hits stable (1.16)

**Comment by swick** [comment]
> There is no priority. --nodevice=fallback:input,all will remove the fallback and the --device=all already present will work.

The implementation doesn't work with the different priorities of the different configurations. You can get device, device fallback and nodevice options from the the runtime, the app, system overrides, system app overrides, user overrides, user app overrides and finally the command line.

At each level the devices get computed and the next level can change that again. Your implementation collects the fallback devices globally and applies them at the end.

**Comment by swick** [comment]
https://github.com/swick/flatpak/commit/19af1b0596ea552d46a523c7fee3ff8f094006ad

This one doesn't flatten anything at all just like the MR but it doesn't touch the `devices` and `devices_valid`  fields. Conditions can be arbitrary things, it generalizes to e.g. sockets and would solve https://github.com/flatpak/flatpak/issues/5789 as well.

App manifest would have `--socket=x11 --socket=wayland --nosocket-if=x11:has-wayland` and a user could force x11 by using `--nosocket-if=x11:false` on the command line.

Here `has_wayland` would be true when flatpak-run sees a usable wayland socket.

Equally, for the usb portal an app could specify `--device=all --nodevice-if=all:can-use-usb-portal` where `can-use-usb-portal` would be true when flatpak-run sees that the portal supports the usb portal.

**Comment by hfiguiere** [comment]
Thing is socket permissions are unrelated. The way fallback-x11 works is because it was a fallback with compatibility mechanism. The bug can be fixed independently.

As for USB and the portal, it's a different PR: #5620

**Comment by swick** [comment]
Sorry, I don't get it. Wayland/X11 fallback is exactly the same as usb to no usb fallback, except that the condition of when the fallback must be active is different (when another socket can be mounted vs when a portal exists). The exact same thing will happen with pulse and pw.

**Comment by hfiguiere** [comment]
It is clear from the conversation that this is not desired.

**Pull request closed without merging**